### PR TITLE
Integrate elastic goodput and slice efficiency into snapshotting

### PR DESCRIPTION
**Pull request by dipannita08**
# Description

This change integrates elastic goodput and slice efficiency into snapshotting.

Following changes are  ported from main inside this PR:

- PR 4840: "Add a new goodput elastic utils function record_slice_state()"
GitHub Commit: [c1572ad0abe6c449aca9e2577003ded11aa07205](https://github.com/AI-Hypercomputer/maxtext/commit/c1572ad0abe6c449aca9e2577003ded11aa07205)

- PR 4846: "Add more elasticity goodput tests" (also contains the elastic_utils fix in the same commit)
GitHub Commit: [cbdde3833e7d86c555212006cd3f3bd39318f051](https://github.com/AI-Hypercomputer/maxtext/commit/cbdde3833e7d86c555212006cd3f3bd39318f051) 

- Unit tests

New changes in this PR (not from main):

- ElasticRecorder API integration into snapshotting version of `train.py`

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/552616060

# Tests

- All simulated tests 72/72 Passed
- Small E2E single host run with elasticity enabled.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/utils/elastic_utils.py [hunk 1]
```diff
@@ -87,13 +87,37 @@ def restore_resharded_state(elastic_mgr: Any, mesh: Any, state: Any):
...
+  active_slices = (
+      active_slices_override if active_slices_override is not None else len(elastic_manager.active_slice_indices)
+  )
+  total_slices = len(elastic.get_slice_to_devices(jax.devices()))
+
+  recorder.record_elastic_slice_counts(
+      available_slices=available_slices,
+      active_slices=active_slices,
+      total_slices=total_slices,
+  )
```
[the remark this answers is not in the document]
Comment (dipannita08) [reply src/maxtext/utils/elastic_utils.py]: Fixed.

**Review by YixuanWang-99 — APPROVED** [review]

**Comment by dipannita08** [comment]
Fixed Pylint issues related to this change, leaving upstream lint issues as is.
